### DISCUSSION TITLE
Using `Arc<AppState>` with `SignedCookieJar`

### DISCUSSION THREAD

**Discussion by woile** [tokio-rs/axum#1556] [Q&A] [answered]
Hello, I've been following some examples that were using `Arc<AppState>` which I was using with `with_state()`. But I just started using `SignedCookieJar` and it creates a problem because I have to implement `impl FromRef<AppState> for Key {...}`.

```
error[E0277]: the trait bound `Key: FromRef<Arc<AppState>>` is not satisfied
   --> src/main.rs:81:31
    |
81  |         .route("/login", post(login))
    |                          ---- ^^^^^ the trait `FromRef<Arc<AppState>>` is not implemented for `Key`
    |                          |
    |                          required by a bound introduced by this call
```

The [example](https://docs.rs/axum-extra/0.4.0-rc.3/axum_extra/extract/struct.SignedCookieJar.html#example) of the `SignedCookieJar` uses directly clone, and I was hoping to avoid it.

Any recommendations?

**Comment by davidpdrsn** [comment, marked as answer]
Since you cannot implement `FromRef<Arc<AppState>> for Key` I'd recommend a newtype:

```rust
#[derive(Clone)]
struct AppState(Arc<InnerState>);

// deref so you can still access the inner fields easily
impl Deref for AppState {
    type Target = InnerState;

    fn deref(&self) -> &Self::Target {
        &*self.0
    }
}

struct InnerState {
    key: Key
}

impl FromRef<AppState> for Key {
    fn from_ref(state: &AppState) -> Self {
        state.0.key.clone()
    }
}
```

**Reply by woile** [reply]
Thanks, this works! Does it make sense to update the SignedCookie to reflect this?

**Reply by davidpdrsn** [reply]
Update it to what?

**Reply by woile** [reply]
I was thinking of adding your snippet or just a few lines [here](https://docs.rs/axum-extra/0.4.0-rc.3/axum_extra/extract/struct.SignedCookieJar.html#example). To give like a recommendation of what to do:

```
If you have been using `Arc<AppState>` you cannot implement `FromRef<Arc<AppState>> for Key`.
The recommendation is to use a new type like: `struct AppState(Arc<InnerState>);` and implement `Deref`.
```

Maybe is something trivial, but when I was following [the axum docs](https://docs.rs/axum/0.6.0-rc.5/axum/index.html#using-the-state-extractor) it uses `Arc` and I was not sure which approach to take.

**Reply by davidpdrsn** [reply]
Sure adding such a note would be good! Do you wanna make a PR?

> Maybe is something trivial, but when I was following [the axum docs](https://docs.rs/axum/0.6.0-rc.5/axum/index.html#using-the-state-extractor) it uses Arc and I was not sure which approach to take.

Yeah we use `Arc` states a bit inconsistently currently.
